### ISSUE TITLE
Terraform doesn't understand pre-release required_version constraint 

### ISSUE THREAD

**Issue by marshall7m** [hashicorp/terraform#28148]
### Terraform Version

```
0.15.0-beta2
```

### Terraform Configuration Files

```
terraform
terraform {
  required_version = "0.15.0-beta2"
}

output "test" {
    value = "foo"
}
```

### Expected Behavior

```
Outputs:

test = "foo"
```
### Actual Behavior

```
│ Error: Unsupported Terraform Core version
│ 
│   on main.tf line 2, in terraform:
│    2:   required_version = "0.15.0-beta2"
│ 
│ This configuration does not support Terraform version 0.15.0-beta2. To proceed, either choose
│ another supported Terraform version or update this version constraint. Version constraints are
│ normally set for good reason, so updating the constraint may lead to other errors or unexpected
│ behavior.
```
### Steps to Reproduce

1. `terraform apply
### Additional Context

Using [tfenv](https://github.com/tfutils/tfenv) for Terraform version management.  Specifically using `tfenv use 0.15.0-beta2` before running `terraform apply`

**Comment by alisdair** [comment]
Hi @marshall7m, thanks for reporting this. I can see this behaviour too.

This is caused by the version package [initializing the `SemVer` variable without the `Prerelease` field](https://github.com/hashicorp/terraform/blob/b9138f44652d9f24fd50f3a5ce4ffbe7b9acd9c7/version/version.go#L27). I think we should be able to update that, but there may be some reason why we don't support constraints against the prerelease field that I'm not aware of.

**Comment by alisdair** [comment]
An update after [a failed attempt to fix this](https://github.com/hashicorp/terraform/tree/alisdair/version-should-include-prerelease): while the issue is at root caused by this mismatched `SemVer` value, this is too load bearing to change. Instead I think we should try updating [the specific location where we check the Terraform version against the constraints](https://github.com/hashicorp/terraform/blob/46c2dce0fb28fe655457fa3e44b85f9bfcccccf5/terraform/version_required.go#L30), and use the prerelease only in that spot.

**Comment by marshall7m** [comment]
thanks @alisdair for taking a look at this. I'm not too familiar with the internal aspects of Terraform (and Go in general) but it seems like your PR will fix it. Thanks!

**Comment by marshall7m** [comment]
I don't know if this helps or brings up another issue but when I use `required_version = "0.15.0"` within all modules used within the target Terraform configuration, Terraform doesn't raise a version. error when I use Terraform version `0.15.0-beta2`.

**Comment by bflad** [comment]
FYI, this also happens when locally installing Terraform CLI. For example, `terraform version` and the version error will output `1.0.11-dev` but the constraint must be `1.0.11`.

**Comment by alisdair** [comment]
I just reinvestigated implementing the second idea I had posted above, and it's also a non-starter. If we consider the full version (including prerelease) when calculating `required_version` compliance, modules with constraints like `>= 0.12.0` will fail with any prerelease version.

This is because [go-version rejects prerelease versions for constraints](https://github.com/hashicorp/go-version/pull/35) which use inequality checks. Making the change I suggested would break the use of modules with any prerelease build.

At the risk of being wrong for the third time in a row, I think the only useful thing we can do here is warn or error if the `required_version` field contains a prerelease version number.
